kv: add a comment about wanting cancellable contexts in TxnCoordSender

andreimatei opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10335)
<!-- Reviewable:end -->

andreimatei committed:
kv: add a comment about wanting cancellable contexts in TxnCoordSender

tschottdorf commented on pkg/kv/txn_coord_sender.go:
```diff
@@ -725,6 +725,10 @@ func (tc *TxnCoordSender) heartbeat(ctx context.Context, txnID uuid.UUID) bool {
+	// TODO(andrei): We'd like all transactions to use a cancelable Context and
+	// get rid of hasAbandoned and txnMeta.lastUpdateNanos. The problem is
+	// internal transactions started in various RPC handlers; we should
+	// investigate using a cancelable context from GRPC.
```
I think this might be a little confusing for everyone not involved in our DM. How about

> We should disallow non-cancelable contexts in the heartbeat goroutine and enforce that our `kv` client cancels the context when it's done. This could be possible a) by enforcing that we only use a local client (i.e. never use the grpc transport; this is mostly true right now except for perhaps some debug commands - in this case) or, if a) isn't possible, by making sure that there's always a context with a reasonable cancellation condition (for example, through the grpc connection plus the retry closure wrapper enforcing proper cancellation on top).

tschottdorf commented:
LGTM

andreimatei commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 731 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVRAGNoCOQDINTsXTeC-r1-731:-KVRH7JzXtIWiILqb9lY:bphmxhj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9eac43388c50b40d348a9d8b5b5d5234cd96f93/pkg/kv/txn_coord_sender.go#L731)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I think this might be a little confusing for everyone not involved in our DM. How about
> 
> 
> 
> > We should disallow non-cancelable contexts in the heartbeat goroutine and enforce that our `kv` client cancels the context when it's done. This could be possible a) by enforcing that we only use a local client (i.e. never use the grpc transport; this is mostly true right now except for perhaps some debug commands - in this case) or, if a) isn't possible, by making sure that there's always a context with a reasonable cancellation condition (for example, through the grpc connection plus the retry closure wrapper enforcing proper cancellation on top).
</details>
i'm a bit confused by what you mean... how could one not use a "local client"? The kv client is always colocated with the `TxnCoordSender`, isn't it? You're saying we have RPCs that pass a transaction proto around through grpc?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 731 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVRAGNoCOQDINTsXTeC-r1-731:-KVRJf9jyh06DfqdUcmM:bv5g6l4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9eac43388c50b40d348a9d8b5b5d5234cd96f93/pkg/kv/txn_coord_sender.go#L731)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> i'm a bit confused by what you mean... how could one not use a "local client"? The kv client is always colocated with the `TxnCoordSender`, isn't it? You're saying we have RPCs that pass a transaction proto around through grpc?
</details>
Anyone using `internal/client/rpc_sender.go` will talk to `TxnCoordSender` over the network. I don't think we do that (perhaps outside of debug) but if we make changes here that don't support that mode of operation, we should remove that sender completely (I have an inkling that dist sql will want something like it again, but would be easier if not).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/kv/txn_coord_sender.go, line 731 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVRAGNoCOQDINTsXTeC-r1-731:-KVRNaiBJS-6yMaNJShQ:b7c0qvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9eac43388c50b40d348a9d8b5b5d5234cd96f93/pkg/kv/txn_coord_sender.go#L731)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Anyone using `internal/client/rpc_sender.go` will talk to `TxnCoordSender` over the network. I don't think we do that (perhaps outside of debug) but if we make changes here that don't support that mode of operation, we should remove that sender completely (I have an inkling that dist sql will want something like it again, but would be easier if not).
</details>
holy molly I didn't know about `rpc_sender.go`. I've marked it as deprecated.
Rewrote this comment as you suggested, but replaced the `or` between a) and b) to an `and` - I think when we do this we should audit all the `db.Txn()` calls and check the cancellation story.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/cli/kv.go, line 54 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVRSDTwNkmDKzzVJpXS:-KVRSDTwNkmDKzzVJpXT:bk084f) ([raw file](https://github.com/cockroachdb/cockroach/blob/c825cea48dd3ad415b449ac20e4df73a3ac22364/pkg/cli/kv.go#L54)):*
> ```Go
> // makeDBClient is DEPRECATED. Don't use it since it doesn't provide a
> // cancelable context for the transactions it starts. See comments in
> // TxnCoordSender.hearbeat() about wanting cancelable contexts for all txns.
> ```

`heartbeat`

---

*[pkg/internal/client/rpc_sender.go, line 39 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVRSFTMFNOheuCMIdcQ:-KVRSFTMFNOheuCMIdcR:bk084f) ([raw file](https://github.com/cockroachdb/cockroach/blob/c825cea48dd3ad415b449ac20e4df73a3ac22364/pkg/internal/client/rpc_sender.go#L39)):*
> ```Go
> // This Sender is DEPRECATED. Don't use it since it doesn't provide a cancelable
> // context for the transactions it starts. See comments in
> // TxnCoordSender.hearbeat() about wanting cancelable contexts for all txns.
> ```

`heartbeat`

---

*[pkg/kv/txn_coord_sender.go, line 731 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVRAGNoCOQDINTsXTeC-r1-731:-KVRSNlRQpJhYggKTTi7:b-u41b79) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9eac43388c50b40d348a9d8b5b5d5234cd96f93/pkg/kv/txn_coord_sender.go#L731)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> holy molly I didn't know about `rpc_sender.go`. I've marked it as deprecated.
> Rewrote this comment as you suggested, but replaced the `or` between a) and b) to an `and` - I think when we do this we should audit all the `db.Txn()` calls and check the cancellation story.
</details>
s/cancelation/cancellation/g (or the other way, but right now you're using both forms in this PR)

```
$ git grep client.NewSender
acceptance/terrafarm/farmer.go: sender, err := client.NewSender(rpcContext, f.Addr(i, base.DefaultPort))
cli/kv.go:      sender, err := client.NewSender(
cmd/zerosum/cluster.go: sender, err := client.NewSender(c.rpcCtx, c.rpcAddr(nodeIdx))
internal/client/client_test.go: sender, err := client.NewSender(rpcContext, addr)
kv/db_test.go:  sender, err := client.NewSender(rpc.NewContext(log.AmbientContext{}, &ctx, nil, stopper), addr)
sql/kv_test.go: sender, err := client.NewSender(
```

Don't think all of these uses are going to go away, though none seem particularly critical. So looks like plan b) it is. Fine for a follow-up, though you could do it in this PR when you're already at it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I've remove the deprecation notice on the RpcSender... Because I'm not sure what the future of that guy is... Not sure if we want to completely get rid of it, or if we'd be fine with it if we move over the TxnCoordSender functionality to the client...

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 731 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVRAGNoCOQDINTsXTeC-r1-731:-KVWJuz2qXayYrYqQsC4:b-tnf8ts) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9eac43388c50b40d348a9d8b5b5d5234cd96f93/pkg/kv/txn_coord_sender.go#L731)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/cancelation/cancellation/g (or the other way, but right now you're using both forms in this PR)
> 
> ```
> $ git grep client.NewSender
> acceptance/terrafarm/farmer.go: sender, err := client.NewSender(rpcContext, f.Addr(i, base.DefaultPort))
> cli/kv.go:      sender, err := client.NewSender(
> cmd/zerosum/cluster.go: sender, err := client.NewSender(c.rpcCtx, c.rpcAddr(nodeIdx))
> internal/client/client_test.go: sender, err := client.NewSender(rpcContext, addr)
> kv/db_test.go:  sender, err := client.NewSender(rpc.NewContext(log.AmbientContext{}, &ctx, nil, stopper), addr)
> sql/kv_test.go: sender, err := client.NewSender(
> ```
> 
> Don't think all of these uses are going to go away, though none seem particularly critical. So looks like plan b) it is. Fine for a follow-up, though you could do it in this PR when you're already at it.
</details>
I've dug some more to see what contexts were used here for the "remote sender", as it wasn't clear to me why those weren't "cancellable contexts". It's because we overwrite the grpc context to `context.TODO()` :). I've modified the comment to state that fact. And I've also changed it to say that we might want to move the TxnCoordSender over to the client. See if you like it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-:-KVWLTTiFWvisGCsE-le:b5znkou)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r3.
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 731 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVWZ4pw5McdUQqPJqKh:-KVWZ4pw5McdUQqPJqKi:b1p1lww) ([raw file](https://github.com/cockroachdb/cockroach/blob/f535fb8ed54019cf52ebc73e2d3b32ddc685a9b3/pkg/kv/txn_coord_sender.go#L731)):*
> ```Go
> 	// TODO(andrei): We should disallow non-cancellable contexts in the heartbeat
> 	// goroutine and enforce that our kv client cancels the context when it's
> 	// done. Unfortunately we have remote clients (roachpb.ExternalClient) that
> ```

this still talks about roachpb.ExternalClient as a culprit of some sort, but as we've discovered, it is not to blame.

Give this entire comment another once-over.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 731 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVWZ4pw5McdUQqPJqKh:-KVWZkBSc1XYVQF4DXbi:b-x0ddpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/f535fb8ed54019cf52ebc73e2d3b32ddc685a9b3/pkg/kv/txn_coord_sender.go#L731)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this still talks about roachpb.ExternalClient as a culprit of some sort, but as we've discovered, it is not to blame.
> 
> Give this entire comment another once-over.
</details>
Well it is to blame, in the sense that it is the reason why we have transactions with non-cancellable context (well, it + the server-side, `DBServer`, also mentioned).
There's other transactions without proper contexts, like those schema changers that @vivekmenezes ran into (and I think he said he'll fix :trollface: ), but I think pointing to the `ExternalClient` is the main point of this comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 739 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVWeJWzNyaayRCZfSwJ:-KVWeJX-xgkw4dPMp_D9:b-fmixek) ([raw file](https://github.com/cockroachdb/cockroach/blob/f535fb8ed54019cf52ebc73e2d3b32ddc685a9b3/pkg/kv/txn_coord_sender.go#L739)):*
> ```Go
> 	// co-locate client.Txn and the TxnCoordSender). At that point, we probably
> 	// don't even need to deal with context cancellation any more; the client will
> 	// be trusted to always send an EndRequest when it's done with a transaction.
> ```

Unless it dies, in which case the connection context should short-circuit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 739 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVWeJWzNyaayRCZfSwJ:-KVWehyPVCFKoBb7CaOQ:b-d93lk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f535fb8ed54019cf52ebc73e2d3b32ddc685a9b3/pkg/kv/txn_coord_sender.go#L739)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Unless it dies, in which case the connection context should short-circuit.
</details>
well if the client is in charge of heartbeating and the client dies, then there's no more heartbeating and the txn is cleaned up that way.
Or in other words, if the client is also the one in charge of cancelling the context (there's no "connection context"), and the client dies, then the context is never cancelled.
Right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 739 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVWeJWzNyaayRCZfSwJ:-KVWjQYcM9FHI278mGbG:b1kkqr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f535fb8ed54019cf52ebc73e2d3b32ddc685a9b3/pkg/kv/txn_coord_sender.go#L739)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> well if the client is in charge of heartbeating and the client dies, then there's no more heartbeating and the txn is cleaned up that way.
> Or in other words, if the client is also the one in charge of cancelling the context (there's no "connection context"), and the client dies, then the context is never cancelled.
> Right?
</details>
Oh, yeah. Sorry, brain fart.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/kv/txn_coord_sender.go, line 731 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335#-KVWZ4pw5McdUQqPJqKh:-KW--Ury1M2P4fiM-puT:b7urira) ([raw file](https://github.com/cockroachdb/cockroach/blob/f535fb8ed54019cf52ebc73e2d3b32ddc685a9b3/pkg/kv/txn_coord_sender.go#L731)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Well it is to blame, in the sense that it is the reason why we have transactions with non-cancellable context (well, it + the server-side, `DBServer`, also mentioned).
> There's other transactions without proper contexts, like those schema changers that @vivekmenezes ran into (and I think he said he'll fix :trollface: ), but I think pointing to the `ExternalClient` is the main point of this comment.
</details>
I've expanded the comment more to also accuse the `Sender` interface of being inadequate wrt these contexts

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10335)*
<!-- Sent from Reviewable.io -->
